### DISCUSSION TITLE
[Q] 使用workbook.xlsx.load读取xls文件出错

### DISCUSSION THREAD

**Discussion by yzkno1** [exceljs/exceljs#2440] [Q&A]
## 💬 Questions and Help
请问，这个如何才能导入xls格式的文件呢

**Comment by Alanscut** [comment]
xls格式文件的导入与xlsx格式是相同的，我在本地验证是ok的，#1256
```
var workbook = new Excel.Workbook();
await workbook.xlsx.readFile('test.xls');
```

**Comment by yzkno1** [comment]
> xls格式文件的导入与xlsx格式是相同的，我在本地验证是ok的，#1256
> 
> ```
> var workbook = new Excel.Workbook();
> await workbook.xlsx.readFile('test.xls');
> ```

我用的是workbook.xlsx.load这个方法就读不出来。。。

**Comment by Alanscut** [comment]
@yzkno1 你是浏览器环境还是Node环境? 这个`workbook.xlsx.load`一般在浏览器环境中使用，需要传入的是`Arraybuffer`类型的数据，你可以参阅 #1168 的示例。

```javascript
  async load(data, options) {
    let buffer;
    if (options && options.base64) {
      buffer = Buffer.from(data.toString(), 'base64');
    } else {
      buffer = data;
    }
}
```

**Comment by 3kinglee** [comment]
![image](https://user-images.githubusercontent.com/14836590/89503824-b0b98c00-d7f9-11ea-9662-60f7698d5bd6.png)

![image](https://user-images.githubusercontent.com/14836590/89504145-29204d00-d7fa-11ea-9054-23811989c8f8.png)

为什么同样是这个项目的成员，一个说可以解析xls，一个说不可以解析xls

**Comment by yzkno1** [comment]
> @yzkno1 你是浏览器环境还是Node环境? 这个`workbook.xlsx.load`一般在浏览器环境中使用，需要传入的是`Arraybuffer`类型的数据，你可以参阅 #1168 的示例。
> 
> ```js
>   async load(data, options) {
>     let buffer;
>     if (options && options.base64) {
>       buffer = Buffer.from(data.toString(), 'base64');
>     } else {
>       buffer = data;
>     }
> }
> ```

浏览器环境，我去试试

**Comment by yzkno1** [comment]
> ![image](https://user-images.githubusercontent.com/14836590/89503824-b0b98c00-d7f9-11ea-9662-60f7698d5bd6.png)
> 
> ![image](https://user-images.githubusercontent.com/14836590/89504145-29204d00-d7fa-11ea-9054-23811989c8f8.png)
> 
> 为什么同样是这个项目的成员，一个说可以解析xls，一个说不可以解析xls

这个问题，当时没解决。就搁置了

**Comment by yzkno1** [comment]
> ![image](https://user-images.githubusercontent.com/14836590/89503824-b0b98c00-d7f9-11ea-9662-60f7698d5bd6.png)
> 
> ![image](https://user-images.githubusercontent.com/14836590/89504145-29204d00-d7fa-11ea-9054-23811989c8f8.png)
> 
> 为什么同样是这个项目的成员，一个说可以解析xls，一个说不可以解析xls

所以到底可不可以呀。。。

**Comment by AuroraTea** [comment]
> xls格式文件的导入与xlsx格式是相同的，我在本地验证是ok的，#1256
> 
> ```
> var workbook = new Excel.Workbook();
> await workbook.xlsx.readFile('test.xls');
> ```

实测node环境也不可以,  你不会是把xlsx文件直接重命名改了后缀吧... @Alanscut 
```node
const excel = require('exceljs')

const workbook = new excel.Workbook()
workbook.xlsx.readFile('test.xls').then(() => {
  console.log(workbook )    // out: all properties undefined or []
})
```

out: 
```node
<ref *1> Workbook {
  category: undefined,
  company: undefined,
  created: undefined,
  description: undefined,
  keywords: undefined,
  manager: undefined,
  modified: undefined,
  properties: undefined,
  calcProperties: undefined,
  _worksheets: [],
  subject: undefined,
  title: undefined,
  views: undefined,
  media: [],
  _definedNames: DefinedNames { matrixMap: {} },
  _xlsx: XLSX { workbook: [Circular *1] },
  creator: undefined,
  lastModifiedBy: undefined,
  lastPrinted: undefined,
  language: undefined,
  revision: undefined,
  contentStatus: undefined,
  _themes: {}
}
```

**Comment by zeromake** [comment]
有相同问题，nodejs 下读取 xls 不成功。

**Comment by zurmokeeper** [comment]
@AuroraTea @zeromake 我在本地是使用node16和exceljs v4.3.0都是可以读到的  readFile 读xls也行 load 读也行，你们的node版本和exceljs版本是多少的，方便发下xls文件和代码示例吗？

**Comment by zeromake** [comment]
[test1.xls](https://github.com/exceljs/exceljs/files/11245516/test1.xls)
> node: 19.8.1
> exceljs: ^4.3.0
> os: macosx,windows 均有出现
> xls: 通过 osx 下的 wps 新建的
``` js
import excel from "exceljs";

async function main() {
  const workbook = new excel.Workbook();
  await workbook.xlsx.readFile(path.join(__dirname, 'test1.xls'));
  console.log(workbook.worksheets.length); // 输出为 0
}
main();
```

**Comment by zurmokeeper** [comment]
> ```js
> await workbook.xlsx.readFile(path.join(__dirname, 'test1.xls'));
> ```

应该是你这个文件本身就是有问题的，你把test1.xls改成test.zip，然后解压下这个文件，解压软件直接就报错了。如果是一个正常的文件，是可以正常解压的。我用你的内容新建了一个test2.xls，是可以正常解压的。

![image](https://user-images.githubusercontent.com/25943604/232382109-07fdce0a-57c6-4f75-a6d9-419655550f3c.png)

exceljs是从这些xml里解析数据的，你的那个test1.xls exceljs去读取的时候没有读到这些文件，所以最后读出来的就是空的，所以怀疑你这个文件本身就是有问题的

**Comment by AuroraTea** [comment]
@zurmokeeper 里面是XML不正说明你这是一个XLSX文件吗...

**Comment by zurmokeeper** [comment]
@Alanscut 我指的你的test1.xls里面的文件不完整，你的是这样的

![1681710388277](https://user-images.githubusercontent.com/25943604/232393655-93742807-7544-4a27-8b77-49e69a94c746.jpg)

而我重新新建以后的test2.xls 里面是这样的

![1681710433289](https://user-images.githubusercontent.com/25943604/232393749-5d163def-edcf-41f1-b204-a62c1d8a9a20.jpg)

缺少了一些文件

**Comment by zurmokeeper** [comment]
@Alanscut 有点尴尬，exceljs读取不到，使用xlsx却可以读到，看来还得好好排查下exceljs是不是哪里还有问题，不知道是不是读取压缩文件的那段代码有问题？

```
const XLSX = require('xlsx');
async function xlsx () {
  const workbook = XLSX.readFile('test1.xls');
  console.log('--->workbook', workbook);
}
```

**Comment by zeromake** [comment]
@zurmokeeper 好像是 xls 某些软件实现的格式，并不是标准 zip，windows 下用 7-Zip-Zstd 显示并不是 zip，我下班切换回 windows 帮忙看看 7-Zip-Zstd 显示是什么压缩格式吧（在用黑苹果）。

**Comment by zeromake** [comment]
@zurmokeeper 7z 识别为 Compound
![图片](https://user-images.githubusercontent.com/8318747/232464099-dcf21227-e238-4484-8387-ee2f01a7cff4.png)
windows wps 下创建的 xls 打开会直接报错，GitHub 不认为是 xls 文件，我打成 zip 上传了。
[test2.zip](https://github.com/exceljs/exceljs/files/11249333/test2.zip)

``` js
import excel from "exceljs";
import path from "path";

async function main() {
  const workbook = new excel.Workbook();
  await workbook.xlsx.readFile(path.join(".", 'test2.xls'));
  console.log(workbook.worksheets.length); // 输出为 0
}
main();
```

```
❯ node .\1.mjs
D:\slots-project\slots-config\node_modules\jszip\lib\zipEntries.js:167
                throw new Error("Can't find end of central directory : is this a zip file ? " +
                      ^

Error: Can't find end of central directory : is this a zip file ? If it is, see https://stuk.github.io/jszip/documentation/howto/read_zip.html
    at ZipEntries.readEndOfCentral (D:\slots-project\slots-config\node_modules\jszip\lib\zipEntries.js:167:23)
    at ZipEntries.load (D:\slots-project\slots-config\node_modules\jszip\lib\zipEntries.js:255:14)
    at D:\slots-project\slots-config\node_modules\jszip\lib\load.js:48:24
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async XLSX.load (D:\slots-project\slots-config\node_modules\exceljs\lib\xlsx\xlsx.js:279:17)
    at async XLSX.readFile (D:\slots-project\slots-config\node_modules\exceljs\lib\xlsx\xlsx.js:55:24)
    at async main (file:///D:/slots-project/slots-config/1.mjs:6:3)

Node.js v19.7.0
```

**Comment by zurmokeeper** [comment]
@yzkno1 @zeromake 找到问题了，exceljs是不支持xls文件的读取的。从我之前的截图可以看出，xls格式的和xlsx格式的里面的文件内容是不一样的，早期版本的Excel格式（如.xls文件）是二进制格式，而新版本的Excel格式（如.xls文件）是基于XML的压缩文件格式。所以从源码也可以看出，读取不了xls格式的。之前的代码用例和我说可以读取xls的，其实都是不经意把xlsx格式的手动改成了xls，只是自己不知道，或者别人把文件发过来的时候已经是xlsx改成xls的了。自己用wps新建一个xls的文件保存，直接就可以重现这个问题了，所以有个作者说的是对的。

> @Alanscut 我指的你的test1.xls里面的文件不完整，你的是这样的
> 
> ![1681710388277](https://user-images.githubusercontent.com/25943604/232393655-93742807-7544-4a27-8b77-49e69a94c746.jpg)
> 
> 而我重新新建以后的test2.xls 里面是这样的
> 
> ![1681710433289](https://user-images.githubusercontent.com/25943604/232393749-5d163def-edcf-41f1-b204-a62c1d8a9a20.jpg)
> 
> 缺少了一些文件

**Comment by zeromake** [comment]
@zurmokeeper 行吧，我的 test2.xls 的 wps 是 2013 版比较老，test1.xls 在 macosx 也是比较老的 wps 创建的，已经放弃 xls 了，都用 xlsx 了。

**Comment by Siemienik** [comment]
> 为什么同样是这个项目的成员，一个说可以解析xls，一个说不可以解析xls

ExcelJS is community of independent contributors. We developed this voluntarily. 

XLSX and XLS it's a diffrent format, so diffrent parsers are needed. Currently Exceljs supports only XLSX. If XLS was open properly it was just a lucky or xlsx under the hood. I would like to add XLS support here, but first I need to found a sponsor or company needs that and want to pay for it.
